### PR TITLE
Issue 2668: Enhancing StreamCut Validation on the client side

### LINKED ISSUE #2668: Incomplete verification in ReaderGroupConfig.verifyStartAndEndStreamCuts()
**Problem description**
When start and end stream cuts are passed to reader group, it verifies that the end cut's highest segment is below the begin cut's and the reverse for the lowest segment. However this is not sufficient to catch all invalid pairs of stream cuts. It could be possible for streamcuts to be overlapping and not ordered with respect to one another, but still have the highest and lowest segments be consistent. 

**Problem location**
ReaderGroupConfig.verifyStartAndEndStreamCuts()

**Suggestions for an improvement**
More thoroughly verify that the ranges are consistent. 
This could be generalized by having streamCuts implement comparable.

### PR DESCRIPTION
**Pull request by huide9**
**Change log description**  
enhancement implementation for stream cut validation

**Purpose of the change**  
Fixes #2668 
Also fixes javadoc build warnings

**What the code does**  

- isStreamCutValid API added for utilize the controller implementation to verify key space integrity of the steram cut and make sure there is no key space overlapping at the same time.
- verifies: for the same segment in different cuts, the position (offset) sequence should be consistent 

**How to verify it**  
CleintFactoryTest - to be done

### CODE REVIEW

**Comment by huide9** [comment]
Need comments from Tom and Shivesh. 
1. code location - moved from ReaderGroupConfig to ClientFactoryImpl
2. the timing to run the validation - need to be explicitly called by others VS. user object creation (in ReaderGroupConfig builder() )
3. if the API added fit for our convention

File: client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java [hunk 1]
```diff
@@ -239,4 +243,31 @@ public void close() {
...
+        final Map<Segment, Long> endPositions = endStreamCut.asImpl().getPositions();
+
+        //* verify both stream cut satisfy the "cover full key space" and "no key space overlap" in each of the streamcuts
+        Preconditions.checkArgument(controller.isStreamCutValid(startStreamCut).equals(true),
+                "startStreamCut key space violation.");
+        Preconditions.checkArgument(controller.isStreamCutValid(endStreamCut).equals(true),
+                "endStreamCut key space violation.");
+
+        //* check if the offsets are legal
+        Preconditions.checkArgument(startPositions.entrySet().stream().allMatch(s ->
```
Comment (shiveshr) [inline client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java]: this may not be sufficient. as suggested earlier, add a new grpc api on controller that takes two streamcuts and responds if start is strictly less than end.

File: client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java [hunk 2]
```diff
@@ -239,4 +243,31 @@ public void close() {
...
+                    }
+                });
+    }
+
+    private void verifyStartAndEndStreamCuts(final StreamCut startStreamCut, final StreamCut endStreamCut) {
+        final Map<Segment, Long> startPositions = startStreamCut.asImpl().getPositions();
+        final Map<Segment, Long> endPositions = endStreamCut.asImpl().getPositions();
+
+        //* verify both stream cut satisfy the "cover full key space" and "no key space overlap" in each of the streamcuts
+        Preconditions.checkArgument(controller.isStreamCutValid(startStreamCut).equals(true),
```
Comment (shiveshr) [inline client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java]: these are remove calls being made to controller. you may want to chain these futures for optimization.
secondly, if a new api is introduced that takes two stream cuts, then the validation of individual cuts can happen as part of that single api call and you dont need to make two separate calls into controller.

Comment (shiveshr) [reply client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java]: also, `controller.isStreamCutValid(endStreamCut)` returns a future.. you cannot compare it with a boolean

File: client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java [hunk 3]
```diff
@@ -239,4 +243,31 @@ public void close() {
         connectionFactory.close();
     }
 
+    public void validateStartAndEndStreamCuts(final ReaderGroupConfig readerGroupConfig) {
+        readerGroupConfig.getEndingStreamCuts().entrySet().stream().filter(e -> !e.getValue().equals(StreamCut.UNBOUNDED))
+                .forEach(e -> {
+                    if (readerGroupConfig.getStartingStreamCuts().get(e.getKey()) != StreamCut.UNBOUNDED) {
```
Comment (andreipaduroiu) [inline client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java]: Cache this value into a local variable. You are essentially fetching it twice.

File: client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java (hunk 2 above)
Comment (andreipaduroiu) [reply client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java]: Correct. There is no need to use `equals(boolean)`; just use ` == true` and the compiler would have caught this problem.

Also, if it's a Future, then this must be making a call over the wire. This will make this method very slow.

File: client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java [hunk 4]
```diff
@@ -239,4 +243,31 @@ public void close() {
...
+                "startStreamCut key space violation.");
+        Preconditions.checkArgument(controller.isStreamCutValid(endStreamCut).equals(true),
+                "endStreamCut key space violation.");
+
+        //* check if the offsets are legal
+        Preconditions.checkArgument(startPositions.entrySet().stream().allMatch(s ->
+                endPositions.entrySet().stream().allMatch(e ->
+                        e.getKey().getSegmentId() == s.getKey().getSegmentId() &&
+                                startPositions.get(s.getKey()) < endPositions.get(e.getKey()))),
+                "Streamcut segment offset illegal.");
```
Comment (andreipaduroiu) [inline client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java]: It may help to include more details in the error message, such as an illegal segment id or offending offsets (including segment id).

File: client/src/main/java/io/pravega/client/stream/impl/ControllerImpl.java [hunk 5]
```diff
@@ -663,6 +664,26 @@ private Boolean handleScaleResponse(Stream stream, ScaleResponse response, long
...
+        final CompletableFuture<StreamCutValidityResponse> result = this.retryConfig.runAsync(() -> {
+            RPCAsyncCallback<StreamCutValidityResponse> callback = new RPCAsyncCallback<>(traceId, "isStreamCutValid");
+            client.isStreamCutValid(io.pravega.controller.stream.api.grpc.v1.Controller.StreamCut.newBuilder().putAllCut(getStreamCutMap(streamCut)).build(),
+                    callback);
+            return callback.getFuture();
+        }, this.executor);
+        return result.thenApply(StreamCutValidityResponse::getResponse)
+                .whenComplete((x, e) -> {
+                    if (e != null) {
+                        log.warn("isStreamCutValid failed: ", e);
```
Comment (andreipaduroiu) [inline client/src/main/java/io/pravega/client/stream/impl/ControllerImpl.java]: Why bother to log this exception here when it would bubble up to the caller anyway?

File: client/src/main/java/io/pravega/client/stream/impl/ReaderGroupState.java [hunk 6]
```diff
@@ -211,6 +211,7 @@ boolean isReaderOnline(String reader) {
 
     /**
      * Returns the number of segments currently being read from and that are unassigned within the reader group.
+     * @return The number of segments currently being read from and that are unassigned within the reader group.
```
Comment (andreipaduroiu) [inline client/src/main/java/io/pravega/client/stream/impl/ReaderGroupState.java]: These Javadoc changes are unrelated to the change at hand. Please undo them and submit them using a separate PR.

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/LocalController.java [hunk 7]
```diff
@@ -373,6 +373,12 @@ private StreamSegments getStreamSegments(List<SegmentRange> ranges) {
         return controller.isSegmentValid(segment.getScope(), segment.getStreamName(), segment.getSegmentId());
     }
 
+    @Override
+    public CompletableFuture<StreamSegments> isStreamCutValid(final StreamCut streamCut) {
+        return controller.isStreamCutValid(streamCut)
+                .thenApply(this::isStreamCutValid);
```
Comment (andreipaduroiu) [inline controller/src/main/java/io/pravega/controller/server/eventProcessor/LocalController.java]: This smells like an infinite loop.

**Comment by andreipaduroiu** [comment]
@huide9 Please check your PR description and title for typos.

File: client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java [hunk 8]
```diff
@@ -239,4 +243,31 @@ public void close() {
         connectionFactory.close();
     }
 
+    public void validateStartAndEndStreamCuts(final ReaderGroupConfig readerGroupConfig) {
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java]: Does this need to be public? Is default sufficient?

Comment (tkaitchuck) [reply client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java]: I don't see where this is called.

File: client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java (hunk 3 above)
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java]: Don't we also need to check that the streams with just a start or just an end have a stream cut which is valid in that it covers all keyspace?

File: client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java (hunk 2 above)
Comment (tkaitchuck) [reply client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java]: Please write a test to cover this case.

File: client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java [hunk 9]
```diff
@@ -239,4 +243,31 @@ public void close() {
...
+    }
+
+    private void verifyStartAndEndStreamCuts(final StreamCut startStreamCut, final StreamCut endStreamCut) {
+        final Map<Segment, Long> startPositions = startStreamCut.asImpl().getPositions();
+        final Map<Segment, Long> endPositions = endStreamCut.asImpl().getPositions();
+
+        //* verify both stream cut satisfy the "cover full key space" and "no key space overlap" in each of the streamcuts
+        Preconditions.checkArgument(controller.isStreamCutValid(startStreamCut).equals(true),
+                "startStreamCut key space violation.");
+        Preconditions.checkArgument(controller.isStreamCutValid(endStreamCut).equals(true),
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java]: Please write a test to cover this case.

File: client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java [hunk 10]
```diff
@@ -239,4 +243,31 @@ public void close() {
...
+                "endStreamCut key space violation.");
+
+        //* check if the offsets are legal
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java]: no *

File: client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java [hunk 11]
```diff
@@ -239,4 +243,31 @@ public void close() {
...
+        Preconditions.checkArgument(startPositions.entrySet().stream().allMatch(s ->
+                endPositions.entrySet().stream().allMatch(e ->
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java]: This looks like it is going to iterate over every end position for each start position. That is going to make this an n^2 computation needlessly.

File: client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java [hunk 12]
```diff
@@ -239,4 +243,31 @@ public void close() {
...
+                        e.getKey().getSegmentId() == s.getKey().getSegmentId() &&
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java]: This does not cover the case where the segment IDs don't match.

**Comment by andreipaduroiu** [comment]
@huide9 What is the status of this PR? I see no activity since February. If we aren't planning on merging it, I suggest we close it.

**Comment by andreipaduroiu** [comment]
Closing for inactivity.
